### PR TITLE
Feature/528 admin role for admin endpoints

### LINKED ISSUE #528: Admin role - for admin endpoint - switch on/off scheduler
Add a role base authorization for admin endpoint - switch on/off scheduler

https://github.com/AbsaOSS/enceladus/pull/1670

### PR DESCRIPTION
**Pull request by jozefbakus**
Implements: #528
New application properties:
`auth.admin.role` - optional - To disable, remove auth.admin.role property
`auth.inmemory.admin.user` - optional -  default value: `hyperdriver-admin-user`
`auth.inmemory.admin.password` - optional - default value `hyperdriver-admin-password`

### CODE REVIEW

File: src/main/scala/za/co/absa/hyperdrive/trigger/configuration/application/AuthConfig.scala [hunk 1]
```diff
@@ -19,6 +19,7 @@ package za.co.absa.hyperdrive.trigger.configuration.application
 import org.springframework.boot.context.properties.bind.{DefaultValue, Name}
 import org.springframework.boot.context.properties.{ConfigurationProperties, ConstructorBinding}
 import org.springframework.validation.annotation.Validated
+import za.co.absa.hyperdrive.trigger.configuration.application.SparkConfig.toNonEmptyOption
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/configuration/application/AuthConfig.scala]: I'd extract this to separate config util or so. Makes no sense to have a dependence to `SparkConfig` from `AuthConfig`

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/auth/InMemoryAuthentication.scala [hunk 2]
```diff
@@ -35,11 +38,15 @@ class InMemoryAuthentication @Inject()(authConfig: AuthConfig) extends Hyperdriv
...
+    auth.inMemoryAuthentication()
       .passwordEncoder(NoOpPasswordEncoder.getInstance())
       .withUser(username)
       .password(password)
       .authorities("ROLE_USER")
+      .and()
+      .passwordEncoder(NoOpPasswordEncoder.getInstance())
+      .withUser(adminUsername)
+      .password(adminPassword)
+      .authorities(adminRole.getOrElse("ROLE_USER"))
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/auth/InMemoryAuthentication.scala]: Please extract `"ROLE_USER"` to a common variable

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/auth/AuthConstants.scala [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import za.co.absa.hyperdrive.trigger.configuration.application.AuthConfig
+
+@Component("authConstants")
+class AuthConstants @Autowired()(authConfig: AuthConfig) {
+  private val adminRole: Option[String] = authConfig.adminRole
+
+  def hasAdminRole(auth: Authentication): Boolean = {
+    adminRole.forall(auth.getAuthorities.toArray(Array[GrantedAuthority]()).map(auth => auth.getAuthority).contains(_))
+  }
+}
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/auth/AuthConstants.scala]: Please add a new-line

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/auth/LdapGrantedAuthoritiesMapper.scala [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  import scala.collection.JavaConversions._
+
+  override def mapAuthorities(grantedAuthorities: util.Collection[_ <: GrantedAuthority]): util.Collection[_ <: GrantedAuthority] = {
+    if (grantedAuthorities == null) {
+      AuthorityUtils.NO_AUTHORITIES
+    } else {
+      grantedAuthorities.map(grantedAuthority => new SimpleGrantedAuthority(grantedAuthority.getAuthority)).toList
+    }
+  }
+}
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/auth/LdapGrantedAuthoritiesMapper.scala]: Please add a new line

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/controllers/RestErrorHandler.scala [hunk 5]
```diff
@@ -38,6 +39,12 @@ class RestErrorHandler {
     new ResponseEntity(HttpStatus.BAD_REQUEST)
   }
 
+  @ExceptionHandler(Array(classOf[AccessDeniedException]))
+  def handleAccessDeniedException(ex: AccessDeniedException, request: WebRequest): ResponseEntity[Object] = {
+    logger.error("Access is denied")
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/controllers/RestErrorHandler.scala]: Did you not pass the `ex` as a 2nd argument on purpose?

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/auth/AuthConstants.scala [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+package za.co.absa.hyperdrive.trigger.api.rest.auth
+
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.security.core.{Authentication, GrantedAuthority}
+import org.springframework.stereotype.Component
+import za.co.absa.hyperdrive.trigger.configuration.application.AuthConfig
+
+@Component("authConstants")
+class AuthConstants @Autowired()(authConfig: AuthConfig) {
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/auth/AuthConstants.scala]: Hmm, it's a bit a strange name `Constants` for a class with methods. Maybe `AuthUtils`?

File: src/main/scala/za/co/absa/hyperdrive/trigger/configuration/application/AuthConfig.scala (hunk 1 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/configuration/application/AuthConfig.scala]: True, I didn't notice the origin of the method. 
Fixed.

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/auth/InMemoryAuthentication.scala (hunk 2 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/auth/InMemoryAuthentication.scala]: Fixed

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/auth/AuthConstants.scala (hunk 3 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/auth/AuthConstants.scala]: Fixed

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/auth/LdapGrantedAuthoritiesMapper.scala (hunk 4 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/auth/LdapGrantedAuthoritiesMapper.scala]: Fixed

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/controllers/RestErrorHandler.scala (hunk 5 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/controllers/RestErrorHandler.scala]: Yes, but I have no argument for it.
I added it to log.

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/auth/AuthConstants.scala (hunk 6 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/auth/AuthConstants.scala]: Yeah true, I've copied it and it originally contained constants.
Done.

**Review by kevinwallimann — APPROVED** [review]

**Pull request merged**